### PR TITLE
Changed bar lines to follow snap size

### PR DESCRIPTION
**Pull request by BoredGuy1**
**Overview**
Vertical lines now appear or disappear depending on quantization value.

Vertical lines between bars are 1 pixel wide with no emboss. Vertical lines at the end of bars are now 2 pixels wide with 1 pixel emboss to make them stand out. (MIN_PIXELS_PER_BAR has been increased from 2 to 4 to compensate.)

To account for small quantization values, vertical lines between bars will automatically disappear as the user zooms out. (Vertical lines at the end of bars will not disappear.)

**Related Threads**
#[5111](https://github.com/LMMS/lmms/issues/5111), #[6420](https://github.com/LMMS/lmms/issues/6420)

**Screenshots**
2 bar quantization (big lines appear every other bar)
2 bar
1/2 bar quantization (big lines appear every bar, one small line appears within each bar)
half bar
1/16 bar quantization (big lines appear every bar, 15 small lines appear within each bar)
16th bar

### CODE REVIEW

**Comment by Rossmaxx** [comment]
Good job. Nice to have this feature implemented but it would help if you could change the defaults too acc to #6420. IMO, 1/16 should be the default snapping with a bit of zoom in. That's what most daws follow these days.

**Comment by BoredGuy1** [comment]
Alright, I'll do that.

**Comment by BoredGuy1** [comment]
Changed the colors a bit to make the coarse grid stand out more. The coarse grid, fine grid, and horizontal line colors can all be individually configured as QSS properties. The line widths are defined as constants in TrackContentWidget.cpp.
image

**Comment by BoredGuy1** [comment]
The coarse grid width, fine grid width, emboss width, and horizontal line width are all configurable now as QSS properties, as well as the offset of the emboss. The default themes have been updated accordingly.

This does break backwards compatibility with old 1.2 themes, but I assumed that that was going to happen sooner or later with 1.3.

image

**Comment by Rossmaxx** [comment]
Requesting review from @michaelgregorius

File: src/gui/tracks/TrackContentWidget.cpp [hunk 1]
```diff
@@ -82,16 +95,27 @@ TrackContentWidget::TrackContentWidget( TrackView * parent ) :
 
 
 
-
-
-
 void TrackContentWidget::updateBackground()
 {
+	// use snapSize to determine number of lines to draw
+	float snapSize = getGUI()->songEditor()->m_editor->getSnapSize();
```
Comment (michaelgregorius) [inline src/gui/tracks/TrackContentWidget.cpp:101]: I'm getting a crash here when I start the application and immediately close it.

File: include/TrackContentWidget.h [hunk 2]
```diff
@@ -51,16 +51,22 @@ class TrackContentWidget : public QWidget, public JournallingObject
 	// qproperties for track background gradients
 	Q_PROPERTY( QBrush darkerColor READ darkerColor WRITE setDarkerColor )
 	Q_PROPERTY( QBrush lighterColor READ lighterColor WRITE setLighterColor )
-	Q_PROPERTY( QBrush gridColor READ gridColor WRITE setGridColor )
+	Q_PROPERTY( QBrush coarseGridColor READ coarseGridColor WRITE setCoarseGridColor )
```
Comment (michaelgregorius) [inline include/TrackContentWidget.h:54]: I hate to be that guy because I usually loathe these comments in reviews of my own code, but can you please unnecessary whitespace between the parentheses? The "rule" is that code that is added should not use these whitespace anymore.

Edit: Please check the diff and remove it everywhere in the new code.

File: src/gui/tracks/TrackContentWidget.cpp [hunk 3]
```diff
@@ -82,16 +95,27 @@ TrackContentWidget::TrackContentWidget( TrackView * parent ) :
...
 	// Assume even-pixels-per-bar. Makes sense, should be like this anyways
 	int ppb = static_cast<int>( tcv->pixelsPerBar() );
 
+	// Coarse grid appears every bar (less frequently if quantization > 1 bar)
+	// Fine grid handles smaller grids
+	float coarseGridResolution = (snapSize >= 1) ? snapSize : 1;
+	float fineGridResolution = snapSize;
+	// Decrease fine grid resolution if it results in less than
+	// 4 pixels between each line
+	while ((ppb * fineGridResolution) < MIN_PIXELS_BETWEEN_LINES)
```
Comment (michaelgregorius) [inline src/gui/tracks/TrackContentWidget.cpp:114]: This can be computed directly instead of converging in a while loop.

File: src/gui/tracks/TrackContentWidget.cpp [hunk 4]
```diff
@@ -101,22 +125,29 @@ void TrackContentWidget::updateBackground()
...
 
 	// draw lines
-	// vertical lines
-	pmp.setPen( QPen( gridColor(), 1 ) );
-	for( float x = 0; x < w * 2; x += ppb )
+	// draw fine grid
+	pmp.setPen( QPen( fineGridColor(), fineGridWidth() ) );
+	for( float x = 0; x < w * 2; x += ppb * fineGridResolution )
+	{
+		pmp.drawLine( QLineF( static_cast<int>( x ), 0.0, static_cast<int>( x ), h ) );
```
Comment (michaelgregorius) [inline src/gui/tracks/TrackContentWidget.cpp:132]: Why are you casting to `int` if you are using `QLineF`? The advantage of `QLineF` is that you can use non-integer coordinates.

**Review by michaelgregorius — CHANGES_REQUESTED** [review]

**Comment by michaelgregorius** [comment]
Hi @BoredGuy1,

thanks for your PR! I really like that the Song Editor now defaults to a more zoomed in view!

I'd like to propose to use a default quantization of 1/4 for the Song Editor though because it makes it look less "busy". It would also coincide with the tempo which is BPM, i.e. 1/4 notes per minute.

Heres a visual comparison of the two options:

SongEditor-1-16
SongEditor-1-4

Using 1/4 is also the default in REAPER to compare against another DAW:

SongEditor-REAPER-Comparison

**Comment by BoredGuy1** [comment]
Hey @michaelgregorius, thanks for the review. And sorry for the late response (been busy).

I agree, default 1/4 definitely looks cleaner than 1/16. That should be an easy change.

File: src/gui/tracks/TrackContentWidget.cpp (hunk 4 above)
Comment (BoredGuy1) [reply src/gui/tracks/TrackContentWidget.cpp:132]: Honestly, I'm not sure. I think I left it there on accident. Good catch - I'll remove it.

File: src/gui/tracks/TrackContentWidget.cpp (hunk 3 above)
Comment (BoredGuy1) [reply src/gui/tracks/TrackContentWidget.cpp:114]: Just curious - is this for the sake of performance, or readability? I came up with something that doesn't use loops (and is probably faster) but might be less readable.

It looks like this, for reference:
```
float fineGridResolution = snapSize;

// Increase fine grid resolution (size between lines) if it results in less than	
// 4 pixels between each line
float pixelsBetweenLines = ppb * snapSize;
if (pixelsBetweenLines < MIN_PIXELS_BETWEEN_LINES) {
	// scaleFactor should be a power of 2
	int scaleFactor = 1 << static_cast<int>( std::ceil( std::log2( MIN_PIXELS_BETWEEN_LINES/pixelsBetweenLines ) ) );
	fineGridResolution *= scaleFactor;
}
```

File: src/gui/tracks/TrackContentWidget.cpp (hunk 1 above)
Comment (BoredGuy1) [reply src/gui/tracks/TrackContentWidget.cpp:101]: Could you describe this issue some more? For some reason, it works fine on my end.

File: include/TrackContentWidget.h (hunk 2 above)
Comment (BoredGuy1) [reply include/TrackContentWidget.h:54]: Haha, yeah I can do that.

Should I remove the whitespaces for the whole file for consistency's sake, or just the lines that I'm adding?

Comment (Rossmaxx) [reply include/TrackContentWidget.h:54]: Just the lines that you touch.

File: src/gui/tracks/TrackContentWidget.cpp (hunk 1 above)
Comment (michaelgregorius) [reply src/gui/tracks/TrackContentWidget.cpp:101]: The steps to reproduce are:
1. Build a debug build of the application.
2. Start the application in a debugger.
3. Close the application.

Before the application is fully quit the debugger breaks at the location linked above. The stack trace is:
```
lmms::gui::TrackContentWidget::updateBackground(lmms::gui::TrackContentWidget * const this) (/home/lmms-build/LMMS/lmms-others/src/gui/tracks/TrackContentWidget.cpp:101)
lmms::gui::TrackContentWidget::qt_static_metacall(QObject * _o, QMetaObject::Call _c, int _id, void ** _a) (/home/lmms-build/LMMS/lmms-others/build/src/lmmsobjs_autogen/DGKZTCOIDE/moc_TrackContentWidget.cpp:115)
libQt5Core.so.5![Unknown/Just-In-Time compiled code] (Unbekannte Quelle:0)
lmms::Model::dataChanged(lmms::Model * const this) (/home/lmms-build/LMMS/lmms-others/build/src/lmmsobjs_autogen/DGKZTCOIDE/moc_Model.cpp:154)
lmms::AutomatableModel::setValue(lmms::AutomatableModel * const this, const float value) (/home/lmms-build/LMMS/lmms-others/src/core/AutomatableModel.cpp:318)
lmms::AutomatableModel::setRange(lmms::AutomatableModel * const this, const float min, const float max, const float step) (/home/lmms-build/LMMS/lmms-others/src/core/AutomatableModel.cpp:424)
lmms::ComboBoxModel::clear(lmms::ComboBoxModel * const this) (/home/lmms-build/LMMS/lmms-others/src/core/ComboBoxModel.cpp:52)
lmms::ComboBoxModel::~ComboBoxModel(lmms::ComboBoxModel * const this) (/home/lmms-build/LMMS/lmms-others/include/ComboBoxModel.h:52)
lmms::ComboBoxModel::~ComboBoxModel(lmms::ComboBoxModel * const this) (/home/lmms-build/LMMS/lmms-others/include/ComboBoxModel.h:53)
libQt5Core.so.5!QObjectPrivate::deleteChildren() (Unbekannte Quelle:0)
libQt5Widgets.so.5!QWidget::~QWidget() (Unbekannte Quelle:0)
lmms::gui::TrackContainerView::~TrackContainerView(lmms::gui::TrackContainerView * const this) (/home/lmms-build/LMMS/lmms-others/src/gui/editors/TrackContainerView.cpp:127)
lmms::gui::SongEditor::~SongEditor(lmms::gui::SongEditor * const this) (/home/lmms-build/LMMS/lmms-others/include/SongEditor.h:68)
lmms::gui::SongEditor::~SongEditor(lmms::gui::SongEditor * const this) (/home/lmms-build/LMMS/lmms-others/include/SongEditor.h:68)
libQt5Core.so.5!QObjectPrivate::deleteChildren() (Unbekannte Quelle:0)
libQt5Widgets.so.5!QWidget::~QWidget() (Unbekannte Quelle:0)
lmms::gui::Editor::~Editor(lmms::gui::Editor * const this) (/home/lmms-build/LMMS/lmms-others/include/Editor.h:84)
lmms::gui::SongEditorWindow::~SongEditorWindow(lmms::gui::SongEditorWindow * const this) (/home/lmms-build/LMMS/lmms-others/include/SongEditor.h:175)
lmms::gui::SongEditorWindow::~SongEditorWindow(lmms::gui::SongEditorWindow * const this) (/home/lmms-build/LMMS/lmms-others/include/SongEditor.h:175)
lmms::gui::MainWindow::~MainWindow(lmms::gui::MainWindow * const this) (/home/lmms-build/LMMS/lmms-others/src/gui/MainWindow.cpp:256)
lmms::gui::MainWindow::~MainWindow(lmms::gui::MainWindow * const this) (/home/lmms-build/LMMS/lmms-others/src/gui/MainWindow.cpp:259)
libQt5Core.so.5!QObject::event(QEvent*) (Unbekannte Quelle:0)
libQt5Widgets.so.5!QApplicationPrivate::notify_helper(QObject*, QEvent*) (Unbekannte Quelle:0)
libQt5Core.so.5!QCoreApplication::notifyInternal2(QObject*, QEvent*) (Unbekannte Quelle:0)
libQt5Core.so.5!QCoreApplicationPrivate::sendPostedEvents(QObject*, int, QThreadData*) (Unbekannte Quelle:0)
libQt5Core.so.5![Unknown/Just-In-Time compiled code] (Unbekannte Quelle:0)
libglib-2.0.so.0!g_main_context_iteration (Unbekannte Quelle:0)
libQt5Core.so.5!QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) (Unbekannte Quelle:0)
libQt5Core.so.5!QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) (Unbekannte Quelle:0)
libQt5Core.so.5!QCoreApplication::exec() (Unbekannte Quelle:0)
main(int argc, char ** argv)
```

As you can see the `SongEditor` is in the process of being deleted but is accessed by the code in `updateBackground` which in turn is triggered by `ComboBoxModel::clear`.

The problem might potentially be fixed by checking if the instances that are used in `TrackContentWidget::updateBackground` are not `nullptr`.

However, I wonder if the call to `clear` in the destructor of `ComboBoxModel` is not a problem as well. Why should the destructor manipulate the model and emit a signal if it is about to be destroyed anyway? I am aware that this was not introduced by your PR but it might be worth discussing this with other people to make the code more stable so that we do not get into the situations seen in the stack trace.

File: src/gui/tracks/TrackContentWidget.cpp (hunk 3 above)
Comment (michaelgregorius) [reply src/gui/tracks/TrackContentWidget.cpp:114]: For me it's a bit of both, performance and readability. If the reference code would be improved by adding a short comment I would prefer it.

File: src/gui/tracks/TrackContentWidget.cpp (hunk 1 above)
Comment (BoredGuy1) [reply src/gui/tracks/TrackContentWidget.cpp:101]: Ah, I see.

The (quick and dirty) way of fixing it is just to go to `TrackContentWidget.cpp`, check if `songEditor == NULL`, and returning if true.

But I did experiment with the destructor in `ComboBoxModel.h`. For what it's worth, simply commenting out `clear` fixes the issue and doesn't seem to adversely affect the program. I will check for leaks with `valgrind` once I get it to work.

Comment (BoredGuy1) [reply src/gui/tracks/TrackContentWidget.cpp:101]: Ah, I see.

The (quick and dirty) way of fixing it is just to go to `TrackContentWidget.cpp`, check if `songEditor == NULL`, and returning if true.

But I did experiment with the destructor in `ComboBoxModel.h`. For what it's worth, simply commenting out `clear` fixes the issue and doesn't seem to adversely affect the program (although I wouldn't take that as a reliable indicator of anything).

I'll ask around.

Comment (BoredGuy1) [reply src/gui/tracks/TrackContentWidget.cpp:101]: Just asked around the Discord, and no one could find a good reason for `clear` being there. So I guess I'll go with the second solution.

**Comment by BoredGuy1** [comment]
OK, everything should be good now.

File: src/gui/tracks/TrackContentWidget.cpp (hunk 1 above)
Comment (michaelgregorius) [reply src/gui/tracks/TrackContentWidget.cpp:101]: Looks good. The crash is gone.

File: include/TrackContentWidget.h (hunk 2 above)
Comment (michaelgregorius) [reply include/TrackContentWidget.h]: There are still some places where whitespace shows up but for me that would be okay. It blends with the existing code and to me functionality is more important than code style (as long as it is not complete chaos).

**Review by michaelgregorius — APPROVED** [review]

File: data/themes/classic/style.css [hunk 5]
```diff
@@ -349,14 +349,24 @@ lmms--gui--TrackView > QWidget {
...
-	qproperty-darkerColor: qlineargradient(x1:0, y1:0, x2:0, y2:1,
-							stop:0 rgb( 50, 50, 50 ), stop:0.33 rgb( 20, 20, 20 ), stop:1 rgb( 15, 15, 15 ) );
-	qproperty-lighterColor: qlineargradient(x1:0, y1:0, x2:0, y2:1,
-							stop:0 rgb( 50, 50, 50 ), stop:0.33 rgb( 40, 40, 40 ), stop:1 rgb( 30, 30, 30 ) );
-	qproperty-gridColor: rgba( 0, 0, 0, 160 );
-	qproperty-embossColor: rgba( 140, 140, 140, 64 );
+	/* colors */
+	qproperty-darkerColor: qlineargradient(x1:0, y1:0, x2:0, y2:1, stop:0 rgb( 50, 50, 50 ), stop:0.33 rgb( 20, 20, 20 ), stop:1 rgb( 15, 15, 15 ) );
+	qproperty-lighterColor: qlineargradient(x1:0, y1:0, x2:0, y2:1, stop:0 rgb( 50, 50, 50 ), stop:0.33 rgb( 40, 40, 40 ), stop:1 rgb( 30, 30, 30 ) );
+	qproperty-coarseGridColor: rgba(0, 0, 0, 160);
```
Comment (zonkmachine) [inline data/themes/classic/style.css:357]: You are mixing styles in this PR. As an example: The line above is according to our convention but the two lines above that is not. Here's the part describing spaces and braces.
https://github.com/LMMS/lmms/wiki/Coding-Conventions#braces

**Review by zonkmachine — CHANGES_REQUESTED** [review]
Just one comment on coding convention.

File: data/themes/classic/style.css (hunk 5 above)
Comment (zonkmachine) [reply data/themes/classic/style.css:357]: OK. I see you have all looked into this previously and there is no need to rant.

File: include/ComboBoxModel.h [hunk 6]
```diff
@@ -47,10 +47,7 @@ class LMMS_EXPORT ComboBoxModel : public IntModel
 	{
 	}
 
-	~ComboBoxModel() override
-	{
-		clear();
-	}
+	~ComboBoxModel() override {}
```
Comment (DomClark) [inline include/ComboBoxModel.h:50]: This destructor could be removed entirely, now that it's empty.

File: include/TrackContentWidget.h [hunk 7]
```diff
@@ -82,17 +88,37 @@ class TrackContentWidget : public QWidget, public JournallingObject
...
+	void setFineGridColor(const QBrush & c);
+	void setHorizontalColor(const QBrush & c);
+	void setEmbossColor(const QBrush & c);
+
+	void setCoarseGridWidth(const int & c);
+	void setFineGridWidth(const int & c);
+	void setHorizontalWidth(const int & c);
+	void setEmbossWidth(const int & c);
+
+	void setEmbossOffset(const int& c);
```
Comment (DomClark) [inline include/TrackContentWidget.h:115]: There's no need to take `int`s by const reference - passing by value makes more sense.

File: src/gui/tracks/TrackContentWidget.cpp [hunk 8]
```diff
@@ -74,6 +77,16 @@ TrackContentWidget::TrackContentWidget( TrackView * parent ) :
...
 
+	// Update background if snap size changes
+	connect( getGUI()->songEditor()->m_editor->snappingModel(),
+			SIGNAL( dataChanged() ),
+			this, SLOT( updateBackground() ) );
+
+	// Also update background if proportional snap is enabled/disabled
+	connect( getGUI()->songEditor()->m_editor,
+			SIGNAL( proportionalSnapChanged() ),
+			this, SLOT( updateBackground() ) );
```
Comment (DomClark) [inline src/gui/tracks/TrackContentWidget.cpp:88]: We prefer to use newer function-pointer-based signals and slots, rather than the old macro-based ones. It would look something like this (not tested):
```cpp
	// Update background if snap size changes
	connect(getGUI()->songEditor()->m_editor->snappingModel(), &Model::dataChanged,
			this, &TrackContentWidget::updateBackground);

	// Also update background if proportional snap is enabled/disabled
	connect(getGUI()->songEditor()->m_editor, &SongEditor::proportionalSnapChanged,
			this, &TrackContentWidget::updateBackground);
```

**Comment by BoredGuy1** [comment]
Ready for review again.

**Comment by michaelgregorius** [comment]
> Ready for review again.

Still looks good to me. @DomClark, @zonkmachine and @Rossmaxx: Is it also good for you? There only seem to be resolved comments now.

**Comment by Rossmaxx** [comment]
Yeah it's fine with me. Tbh 1/16 was too cluttered to begin with. Thanks to Michael to suggest the 1/4 snapping. I didn't know that 1/4 was supposed to be better. 

LGTM

**Review by DomClark — APPROVED** [review]
